### DISCUSSION TITLE
Merge to support union schemas

### DISCUSSION THREAD

**Discussion by indifferentghost** [open-circle/valibot#299] [Q&A] [answered]
Hey,

I'm doing some validation on environment variables. Attempting something like the following:
```ts
const DevSchema = object({
  NODE_ENV: literal('production'),
  DATABASE_URL: string([toTrimmed(), minLength(1), url(), toLowerCase(), excludes('?tls=0')])
});

const ProdSchema = object({
  NODE_ENV: literal('development'),
  DATABASE_URL: string([toTrimmed(), minLength(1), url(), toLowerCase(), endsWith('?tls=0')])
});

const EnvSchema = merge([
  union([ DevSchema, ProdSchema ]),
  object({
    // shared config
    FOO_BAR: string()
  }),
])
```
Throws a lengthy TypeScript error that boils down to merge not supporting union schemas (and that's [shown in the types](https://github.com/fabian-hiller/valibot/blob/main/library/src/methods/merge/merge.ts#L13-L17) right now.)

I looked and while I [found references of it being implemented in 0.20.0](https://github.com/fabian-hiller/valibot/issues/90#issuecomment-1786480237).

I still unsure even what code to reference or how to build it.

**Comment by indifferentghost** [comment]
I ended up forcing `descriminatedUnion` the deprecation notice put me to `variant`, but I'm still unsure on the API of how to use this:

```ts
const EnvSchema = merge([
	object({ AUTH_SIGNED_COOKIE: string([toTrimmed(), minLength(1)]) }),
	variant('NODE_ENV', [DevEnvSchema, ProdEnvSchema])
	// union([DevEnvSchema, ProdEnvSchema]),
]);
```

**Reply by fabian-hiller** [reply]
As you can see [here](https://github.com/fabian-hiller/valibot/blob/v0.24.1/library/src/schemas/variant/variant.ts#L155), `descriminatedUnion` is equal to `variant`. So please use `variant` instead. The idea of `variant` is to define the discriminator as the first argument and the union options as the second one. This way the validation is faster and the issues are more specific.

**Reply by fabian-hiller** [reply]
`variant` is basically a specialised version of `union` for objects.

**Comment by fabian-hiller** [comment, marked as answer]
Please try to use `intersect` for this use case. `merge` does only work for pure objects because it is able to truly merge objects together. For example, `merge` is able to overwrite individual entries.

```ts
const EnvSchema = intersect([
  union([DevSchema, ProdSchema]),
  object({
    // shared config
    FOO_BAR: string()
  }),
]);
```

**Reply by indifferentghost** [reply]
Works brilliantly! Thanks!

**Reply by indifferentghost** [reply]
Is there a reason to use a union over a variant in this senario?
```ts
const EnvSchema = intersect([
	object({ AUTH_SIGNED_COOKIE: string([toTrimmed(), minLength(1)]) }),
	variant('NODE_ENV', [DevEnvSchema, ProdEnvSchema])
]);
```

**Reply by fabian-hiller** [reply]
Whenever you work with objects and it is possible to use `variant` you should prefer it over `union`.

**Reply by fabian-hiller** [reply]
The reason is that the validation is faster and the output in case of any issues is more specific. Therefore, it is easier for you or your end users to fix the dataset.

**Reply by indifferentghost** [reply]
> Whenever you work with objects and it is possible to use `variant` you should prefer it over `union`.

This is gold thank you
